File: arcusApp/app/src/main/java/arcus/app/device/DeviceListingFragment.java
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.device;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import com.iris.client.model.DeviceModel;
import com.iris.client.model.PlaceModel;
import arcus.app.R;
import arcus.app.common.backstack.BackstackManager;
import arcus.app.common.controller.BackstackPopListener;
import arcus.app.common.fragments.BaseFragment;
import arcus.app.common.image.ImageManager;
import arcus.app.common.image.Wallpaper;
import arcus.app.common.models.RegistrationContext;
import arcus.app.common.models.SessionModelManager;
import arcus.app.device.adapter.DeviceListAdapter;
import arcus.app.device.details.DeviceDetailParentFragment;
import arcus.app.device.zwtools.controller.ZWaveToolsSequence;

import java.util.ArrayList;
import java.util.List;

public class DeviceListingFragment extends BaseFragment implements BackstackPopListener, SessionModelManager.SessionModelChangeListener {
    private List<DeviceModel> devicesPaired = new ArrayList<>();
    private DeviceListAdapter deviceListAdapter;
    private ListView deviceListView;
    private TextView numOfDevicesTextView;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View mView = super.onCreateView(inflater, container, savedInstanceState);
        numOfDevicesTextView = (TextView) mView.findViewById(R.id.tvNumOfDevices);
        deviceListView = (ListView) mView.findViewById(R.id.lvDeviceList);

        return (mView);
    }

    @Override
    public void onResume() {
        super.onResume();
        getActivity().setTitle(getTitle());

        deviceListAdapter = new DeviceListAdapter(getActivity(), devicesPaired);
        deviceListView.setAdapter(deviceListAdapter);
        deviceListAdapter.notifyDataSetChanged();

        if(numOfDevicesTextView != null && devicesPaired != null){
            String s = devicesPaired.size()+"";
            numOfDevicesTextView.setText(s);
        }

        deviceListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapter, View v, int position, long arg3) {
                removeListener();
                if (deviceListAdapter.getItemViewType(position) == DeviceListAdapter.TYPE_ITEM) {
                    BackstackManager.getInstance()
                            .navigateToFragment(DeviceDetailParentFragment.newInstance(position), true);
                } else {
                    new ZWaveToolsSequence().startSequence(getActivity(), null);
                }
            }
        });

        if (!SessionModelManager.instance().isListenerRegistered(this)) {
            addListener();
        }
        getDevicesPaired();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        removeListener();
        if (SessionModelManager.instance().isListenerRegistered(this))
            removeListener();
    }

    @Override
    public void onPopped() {
        String placeID = "";
        PlaceModel model = RegistrationContext.getInstance().getPlaceModel();
        if (model != null) {
            placeID = model.getId();
        }
        ImageManager.with(getActivity())
              .setWallpaper(Wallpaper.ofPlace(placeID).darkened());
    }

    @NonNull
    @Override
    public String getTitle() {
        return "Devices";
    }

    @Override
    public Integer getLayoutId() {
        return R.layout.fragment_device_listing;
    }

    @Override
    public Integer getMenuId() {
        return R.menu.menu_device_list;
    }

    private void addListener() {
        SessionModelManager.instance().addSessionModelChangeListener(this);
    }

    private void removeListener() {
        SessionModelManager.instance().removeSessionModelChangeListener(this);
    }

    @Override
    public void onSessionModelChangeEvent(SessionModelManager.SessionModelChangedEvent event) {
        getDevicesPaired();
    }

    private void getDevicesPaired() {
        devicesPaired = SessionModelManager.instance().getDevicesWithHub();
        if (devicesPaired != null) {
            deviceListAdapter.setDevices(devicesPaired);
            numOfDevicesTextView.setText(String.format("%d", devicesPaired.size()));
        }
    }
}

File: app/src/main/java/arcus/app/device/list/DeviceListingFragment.kt
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.device.list

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import arcus.app.R
import arcus.app.common.backstack.BackstackManager
import arcus.app.common.fragments.CoreFragment
import arcus.app.device.details.DeviceDetailParentFragment
import arcus.app.device.list.DeviceListAdapter.ItemClickListener
import arcus.app.device.zwtools.controller.ZWaveToolsSequence
import arcus.presentation.common.view.ViewState
import arcus.presentation.device.list.DeviceListViewModel

class DeviceListingFragment : CoreFragment<DeviceListViewModel>() {
    private lateinit var deviceListAdapter: DeviceListAdapter
    private lateinit var numOfDevices: TextView

    override val title: String get() = getString(R.string.sidenav_devices_title)
    override val layoutId: Int = R.layout.fragment_device_listing
    override val viewModelClass: Class<DeviceListViewModel> = DeviceListViewModel::class.java

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        numOfDevices = view.findViewById(R.id.numberOfDevices)

        val deviceList: RecyclerView = view.findViewById(R.id.deviceList)
        deviceListAdapter = DeviceListAdapter(object : ItemClickListener {
            override fun itemClicked(position: Int) {
                BackstackManager
                        .getInstance()
                        .navigateToFragment(
                                DeviceDetailParentFragment.newInstance(position),
                                true
                        )
            }

            override fun footerClicked() {
                ZWaveToolsSequence().startSequence(activity, null)
            }
        })
        deviceList.adapter = deviceListAdapter
        viewModel.viewState.observe(viewLifecycleOwner, Observer {
            when (it) {
                is ViewState.Loaded -> {
                    val (devices, deviceListItems) = it.item
                    numOfDevices.text = devices.toString()
                    deviceListAdapter.submitList(deviceListItems)
                }
            }
        })
    }
}
